Restore progress reporting

pranavgaikwad opened the pull request:
This PR introduces progress reporting for Velero Restore. 

Closes #21 

The changes in this PR follow the design doc : https://github.com/vmware-tanzu/velero/blob/main/design/restore-progress.md

Summary of changes :
* **pkg/apis/velero/v1/restore.go** : Introduces new field _RestoreProgress_ in _RestoreStatus_
* **config/crd/bases/velero.io_restores.yaml** & **config/crd/crds/crds.go**: Updates Restore CRD to include above field.
* **pkg/cmd/server/server.go** : Passes velero client when creating a new Restore. The client is needed to send updates to _RestoreStatus_ during an ongoing restore
* **pkg/restore/restore.go**: Modifies restore flow as per https://github.com/vmware-tanzu/velero/blob/main/design/restore-progress.md#modifications-to-restorego.
* **pkg/cmd/util/output/restore_describer.go**: Updates `velero describe` command to show restore progress information
* **pkg/restore/restore_test.go**: Adds missing velero client in one of the Restore tests
* **pkg/util/kube/utils.go**: Updates `EnsureNamespaceExistsAndIsReady` method to return an additional boolean to indicate whether a new Namespace was created or not. If new ns created, it will be added to list of restored items. 
* **pkg/util/kube/utils_test.go**: Updates test to verify value of the new boolean

jenting reviewed (changes requested):
1. Could we add the restore progress reporting at `velero restore describe`, same as the backup progress report at `velero backup describe`, something like:
   ```
   Total items to be restored from:
   Items restored from:
   ```

2. I create a backup resource, at backup reporting, the total items are 106. However, when I restore from the backup resource, the total times are 22, why? :thinking:

pranavgaikwad commented:
@jenting Thanks for pointing that out. I will take a look, my first guess is that maybe those resources were skipped during restore. I will confirm though.

pranavgaikwad commented:
@jenting In my environment, I didn't see a huge difference in the number of items backed up vs restored:

```yml
apiVersion: velero.io/v1
kind: Backup
metadata:
  [...]
  labels:
    velero.io/storage-location: default
  name: bkp-1
  namespace: velero
  [...]
spec:
  [...]
  includedNamespaces:
  - test-bkp
  - test-bkp-2
status:
  [...]
  phase: Completed
  progress:
    itemsBackedUp: 389
    totalItems: 389
  startTimestamp: "2021-01-22T19:13:06Z"
```

```yml
apiVersion: velero.io/v1
kind: Restore
metadata:
  name: bkp-1-20210122141607
  namespace: velero
  [...]
spec:
  backupName: bkp-1
  excludedResources:
  - nodes
  - events
  - events.events.k8s.io
  - backups.velero.io
  - restores.velero.io
  - resticrepositories.velero.io
  hooks: {}
  includedNamespaces:
  - '*'
status:
  completionTimestamp: "2021-01-22T19:16:29Z"
  phase: Completed
  progress:
    itemsRestored: 387
    totalItems: 387
  startTimestamp: "2021-01-22T19:16:08Z"
```

The difference of 2 items in Backup and Restore comes from the fact that this PR doesn't include namespaces in the count of items restored. There is a way to update [this function](https://github.com/vmware-tanzu/velero/blob/a42284ed176ca1de2b5e8b9f52dac7b1d1785bb4/pkg/util/kube/utils.go#L48) to return the number of ns it creates so that ns'es can be included in the count of items. 

I'd be interested to know what kind of resources you tried to restore when you saw this huge mismatch in numbers.

jenting commented:
I forgot the original test steps but I found a new one.

Steps:
1. Create a new namespace called `nginx`
2. Install nginx `kubectl apply -f https://k8s.io/examples/application/deployment.yaml`
3. Backup namespace `nginx` `velero backup create backup-nginx --include-namespaces=nginx`
4. Describe backup `velero backup describe backup-nginx --details`
```shell
Total items to be backed up:  31
Items backed up:              31

Resource List:
  apps/v1/Deployment:
    - nginx/nginx-deployment
  apps/v1/ReplicaSet:
    - nginx/nginx-deployment-66b6c48dd5
  v1/Event:
    - nginx/nginx-deployment-66b6c48dd5-9l4n9.165dccc49aae8161
    - nginx/nginx-deployment-66b6c48dd5-9l4n9.165dccc4b5fe1f95
    - nginx/nginx-deployment-66b6c48dd5-9l4n9.165dccc4bac83caf
    - nginx/nginx-deployment-66b6c48dd5-9l4n9.165dccc4c2609ca9
    - nginx/nginx-deployment-66b6c48dd5-9l4n9.165dcccc5f50a571
    - nginx/nginx-deployment-66b6c48dd5-czxj5.165dccd10a7cede5
    - nginx/nginx-deployment-66b6c48dd5-czxj5.165dccd126e35f36
    - nginx/nginx-deployment-66b6c48dd5-czxj5.165dccd13023360e
    - nginx/nginx-deployment-66b6c48dd5-czxj5.165dccd13700b6cf
    - nginx/nginx-deployment-66b6c48dd5-kwwmn.165dccd1083e8e3f
    - nginx/nginx-deployment-66b6c48dd5-kwwmn.165dccd12510c30b
    - nginx/nginx-deployment-66b6c48dd5-kwwmn.165dccd130b5c888
    - nginx/nginx-deployment-66b6c48dd5-kwwmn.165dccd136ed2aa2
    - nginx/nginx-deployment-66b6c48dd5-ph7tc.165dccc49ab2bd34
    - nginx/nginx-deployment-66b6c48dd5-ph7tc.165dccc4b9fc9cf2
    - nginx/nginx-deployment-66b6c48dd5-ph7tc.165dccc4c1536718
    - nginx/nginx-deployment-66b6c48dd5-ph7tc.165dccc4c8c0fc9a
    - nginx/nginx-deployment-66b6c48dd5-ph7tc.165dcccc5ec2d1b3
    - nginx/nginx-deployment-66b6c48dd5.165dccc49802b570
    - nginx/nginx-deployment-66b6c48dd5.165dccc4992f8858
    - nginx/nginx-deployment-66b6c48dd5.165dccd107609102
    - nginx/nginx-deployment-66b6c48dd5.165dccd10865dc6d
    - nginx/nginx-deployment.165dccc4964f2922
    - nginx/nginx-deployment.165dccd10652f80c
  v1/Namespace:
    - nginx
  v1/Pod:
    - nginx/nginx-deployment-66b6c48dd5-czxj5
    - nginx/nginx-deployment-66b6c48dd5-kwwmn
  v1/Secret:
    - nginx/default-token-qwsgw
  v1/ServiceAccount:
    - nginx/default
```
5. Delete namespace `nginx`
6. Restore from backup backup-nginx, `velero restore create restore-nginx --from-backup=backup-nginx --include-namespaces=nginx`
7. Check size `kubectl get restores restore-nginx -o yaml`
```shell
spec:
  backupName: backup-nginx
  excludedResources:
  - nodes
  - events
  - events.events.k8s.io
  - backups.velero.io
  - restores.velero.io
  - resticrepositories.velero.io
  hooks: {}
  includedNamespaces:
  - nginx
status:
  completionTimestamp: "2021-01-26T14:06:10Z"
  phase: Completed
  progress:
    itemsRestored: 6
    totalItems: 6
  startTimestamp: "2021-01-26T14:06:09Z"
```

pranavgaikwad commented:
@jenting As the events are excluded from the restore. All the v1/Event resources present in the backup were not restored and hence the difference in the number.

jenting commented:
> @jenting As the events are excluded from the restore, 24 instances of the v1/Event resource present in the backup were not restored and hence the difference in the number. Please also note that I am working on a way to include namespaces in the count.

Cool, thank you for the explanation.

I'd love that the `velero backup describe` displays the counts on total items to be restored and current restored items.

nrb commented on pkg/restore/restore.go:
```diff
@@ -373,103 +381,89 @@ func (ctx *restoreContext) execute() (Result, Result) {
-			if target, ok := ctx.restore.Spec.NamespaceMapping[namespace]; ok {
-				targetNamespace = target
-			}
+	totalItems, i, existingNamespaces := 0, 0, sets.NewString()
```
What does `i` represent here? An iteration counter?

nrb commented on pkg/restore/restore.go:
```diff
@@ -373,103 +381,89 @@ func (ctx *restoreContext) execute() (Result, Result) {
-				// have to try to create them multiple times
-				existingNamespaces.Insert(targetNamespace)
+				w, e := ctx.restoreItem(obj, groupResource, selectedItem.targetNamespace)
+				actualTotalItems := len(ctx.restoredItems) + (totalItems - (i + 1))
```
I'm unclear as to why we're doing `(totalItems - (i+1)`) here. Can you clarify with a comment?

pranavgaikwad replied:
@nrb since there may be additional items restored by the plugins, the initial count `totalItems` needs to be adjusted to include those additional items. `ctx.restoredItems` has all the items actually restored. But we don't want to count items twice. Therefore, this adjustment is required.

nrb replied:
Ahh, ok. Can you add that as a code comment, please?

nrb commented on pkg/restore/restore.go:
```diff
@@ -1468,3 +1471,140 @@ func isCompleted(obj *unstructured.Unstructured, groupResource schema.GroupResou
+	name            string
+}
+
+func (ctx *restoreContext) getOrderedResourceCollection(backupResources map[string]*archive.ResourceItems) (restoreResourceCollection []restoreableResource, warnings Result, errs Result) {
```
Our current code style is to not use named return variables. Could you move them inline with the function please?

nrb commented on pkg/restore/restore.go:
```diff
@@ -1468,3 +1471,140 @@ func isCompleted(obj *unstructured.Unstructured, groupResource schema.GroupResou
+	return
+}
+
+func (ctx *restoreContext) getSelectedRestoreableItems(resource, targetNamespace, originalNamespace string, items []string) (res restoreableResource, warnings Result, errs Result) {
```
Similar to above function, could you please use the return variables inline?

nrb commented on pkg/restore/restore.go:
```diff
@@ -1468,3 +1471,140 @@ func isCompleted(obj *unstructured.Unstructured, groupResource schema.GroupResou
 	return false, nil
 }
+
+type restoreableResource struct {
```
Can you provide comments on what the differences are between a restoreableResource and a restoreableItem are here?

pranavgaikwad replied:
added comment. can reword if this one sounds confusing.

nrb commented on pkg/restore/restore.go:
```diff
@@ -1468,3 +1471,140 @@ func isCompleted(obj *unstructured.Unstructured, groupResource schema.GroupResou
+	return
+}
+
+func (ctx *restoreContext) getSelectedRestoreableItems(resource, targetNamespace, originalNamespace string, items []string) (res restoreableResource, warnings Result, errs Result) {
```
If I'm reading this code correctly, `getOrderedResourceCollection` is very similar to the previous restore loop in that it's applying the selection criteria like namespace, and includes/excludes. However, the change here is that this function, `getSelectedRestoreableItems` then groups those resources by namespace and applies the Kubernetes label selection. It replaces some of the functionality of the `restoreResource` function, which should be removed since it's no longer relevant with this PR, from what I can tell.

Do I have that right? If so please add a comment to clarify that's what this function is doing, so we can keep track of it.

Also, based on this code it looks like we're unmarshaling twice.

pranavgaikwad replied:
@nrb that is correct. We are unmarshling twice. Unmarshling is needed at first to apply selection criteria, and then finally to actually restore the object. Another alternative was to keep entire object in-memory until it's restored which isnt very attractive.

pranavgaikwad replied:
and your'e correct about the duplicate code from restoreResource function. thanks for pointing that out. will clean it up

nrb replied:
> Another alternative was to keep entire object in-memory until it's restored which isnt very attractive.

Right, that is the trade we'd have to make. I honestly think we will end up doing this at some point, though, as we'd like to form a graph of objects. For now I think this will work and we can revisit the unmarshalling mechanisms in the future.

nrb commented on changelogs/unreleased/3125-pranavgaikwad:
```diff
@@ -0,0 +1 @@
+Restore progress reporting
```
I'd suggest something like "Progress reporting added for restores" for a changelog entry.

nrb reviewed (changes requested):
I think I understand how this is all working, but I'd like some more comments, as the restore code can be a little tricky :)

Overall I think it makes sense, though, and I appreciate your efforts here!

And finally, it needs a rebase before we'll be able to merge.

pranavgaikwad commented:
@jenting This commit addresses your feedback about `velero describe restore` command: https://github.com/vmware-tanzu/velero/pull/3125/commits/35f8e270478e0fad8e55d7be31692801a63043f1

pranavgaikwad commented:
@nrb when you have a moment, request you to take a look at new code I added to include namespaces in the total count. That change updates `EnsureNamespaceExistsAndIsReady()` util method to return whether it creates a new ns or not. If a new ns is created, it is added to the list of restored items.

nrb commented on pkg/util/kube/utils.go:
```diff
@@ -45,7 +45,10 @@ func NamespaceAndName(objMeta metav1.Object) string {
-func EnsureNamespaceExistsAndIsReady(namespace *corev1api.Namespace, client corev1client.NamespaceInterface, timeout time.Duration) (bool, error) {
+func EnsureNamespaceExistsAndIsReady(namespace *corev1api.Namespace, client corev1client.NamespaceInterface, timeout time.Duration) (bool, bool, error) {
+	// nsCreated tells whether the namespace was created by this method
+	// required for keeping track of number of restored items
```
L43 above should be updated to match this, too.

nrb reviewed (changes requested):
Thanks for these updates! The comments are helpful.

I think the new variable tracking whether or not a namespace was actually created is useful, just need to make sure the function comment matches the code.

nrb commented:
@pranavgaikwad Looks like there was a conflict with the colorized output PR that's affecting this.

pranavgaikwad commented:
> @pranavgaikwad Looks like there was a conflict with the colorized output PR that's affecting this.

Sure will rebase again.

jenting commented on pkg/restore/restore.go:
```diff
@@ -373,103 +381,104 @@ func (ctx *restoreContext) execute() (Result, Result) {
+				// it in order to ensure it exists. Try to get it from the backup tarball
+				// (in order to get any backed-up metadata), but if we don't find it there,
+				// create a blank one.
+				if namespace != "" && !existingNamespaces.Has(selectedItem.targetNamespace) {
+					logger := ctx.log.WithField("namespace", namespace)
+					ns := getNamespace(logger, archive.GetItemFilePath(ctx.restoreDir, "namespaces", "", namespace), selectedItem.targetNamespace)
+					if _, nsCreated, err := kube.EnsureNamespaceExistsAndIsReady(ns, ctx.namespaceClient, ctx.resourceTerminatingTimeout); err != nil {
+						errs.AddVeleroError(err)
+						continue
+					} else {
+						// add the newly created namespace to the list of restored items
+						if nsCreated {
+							itemKey := velero.ResourceIdentifier{
+								GroupResource: kuberesource.Namespaces,
+								Namespace:     ns.Namespace,
+								Name:          ns.Name,
+							}
+							ctx.restoredItems[itemKey] = struct{}{}
+						}
+					}
```
This block code looks similar to https://github.com/vmware-tanzu/velero/blob/b527a8a32deb9f5546327314b8347495f7855dc5/pkg/restore/restore.go#L794-L811

And the code logic would run twice. I'm not sure what's the difference between these two or could we merge them into one?

pranavgaikwad replied:
@jenting This duplication is required because `restoreItem()` can be called for additional items returned by plugins. The recursive call in that case only triggers restoreItem() function alone and not the rest of the previous logic.

jenting reviewed (dismissed):
LGTM

jenting commented:
@pranavgaikwad I just realized there is a conflict file. Please resolve it if you have time. 🙇

carlisia commented:
I'll be reviewing this tomorrow.

jenting reviewed (approved):
LGTM, thank you @pranavgaikwad

nrb reviewed (approved):
🎉

rishi-anand commented:
@pranavgaikwad, @jenting, @nrb, @carlisia I have seen this issue again in my cluster which is on `v1.6.0`

I have restored in same cluster with `preserveNodePort: true` and velero was throwing error msg about nodePort already in use. Later, I changed `preserveNodePort: false` but then restore crd says all items restored but phase is `InProgress` from a very long time. I tried restarting velero pod but that didn't fixed the issue.

```
apiVersion: velero.io/v1
kind: Restore
metadata:
  creationTimestamp: "2021-06-08T01:57:44Z"
  name: backup-1
  namespace: velero
  resourceVersion: "13816289"
  selfLink: /apis/velero.io/v1/namespaces/velero/restores/backup-1
  uid: c8cc70b3-4b7a-4bde-aadb-da5856aeb259
spec:
  backupName: backup-1
  excludedResources:
  - nodes
  - events
  - events.events.k8s.io
  - backups.velero.io
  - restores.velero.io
  - resticrepositories.velero.io
  hooks: {}
  includeClusterResources: true
  preserveNodePorts: true
  restorePVs: true
status:
  phase: InProgress
  progress:
    itemsRestored: 1037
    totalItems: 1037
  startTimestamp: "2021-06-08T01:57:44Z"
```
